core/src/main/java/jenkins/util/SystemProperties.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 Johannes Ernst http://upon2020.com/
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.util;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.servlet.ServletContext;
+
+/**
+ * Centralizes calls to <code>System.getProperty()</code> and related calls.
+ * This allows us to get values not just from environment variables but also from
+ * the <code>ServletContext</code>, so properties like <code>hudson.DNSMultiCast.disabled</code>
+ * can be set in <code>context.xml</code> and the app server's boot script does not
+ * have to be changed.
+ * 
+ * <p>While it looks like it on first glance, this cannot be mapped to <code>EnvVars.java</code>
+ * because <code>EnvVars.java</code> is only for build variables, not Jenkins itself variables. 
+ * 
+ * <p>This should be used to obtain hudson/jenkins "app"-level parameters
+ * (e.g. <code>hudson.DNSMultiCast.disabled</code>), but not for system parameters
+ * (e.g. <code>os.name</code>).
+ * 
+ * @author Johannes Ernst
+ * @since 1.639
+ */
+public class SystemProperties {
+    /**
+     * The ServletContext to get the "init" parameters from.
+     */
+    private static ServletContext theContext;
```

Comment:
.. and the same ClassLoaders for all of them? In which case, this should work as before as long as you don't specify properties in context.xml. In case of Tomcat, for example, different ClassLoaders should keep this from being a global singleton.